library/std/src/path.rs

```diff
@@ -2873,9 +2886,35 @@ impl cmp::PartialEq for Path {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl Hash for Path {
     fn hash<H: Hasher>(&self, h: &mut H) {
-        for component in self.components() {
-            component.hash(h);
+        let bytes = self.as_u8_slice();
+
+        let mut component_start = 0;
+        let mut bytes_hashed = 0;
+
+        for i in 0..bytes.len() {
+            if is_sep_byte(bytes[i]) {
+                if i > component_start {
+                    let to_hash = &bytes[component_start..i];
+                    h.write(to_hash);
+                    bytes_hashed += to_hash.len();
+                }
+
+                // skip over separator and optionally a following CurDir item
+                // since components() would normalize these away
+                component_start = i + match bytes[i..] {
+                    [_, b'.', b'/', ..] | [_, b'.'] => 2,
+                    _ => 1,
+                };
+            }
         }
+
+        if component_start < bytes.len() {
+            let to_hash = &bytes[component_start..];
+            h.write(to_hash);
+            bytes_hashed += to_hash.len();
+        }
+
+        h.write_usize(bytes_hashed);
```

Comment:
This seems OK -- I'm not immediately seeing anything wrong -- but trying to map this code as being equivalent to the component iterator equality seems quite difficult. Is there something we can do to build confidence it's correct, particularly as we potentially change the implementation of components()?

Perhaps we could instead improve the performance of next() on Components for similar effect?

Reply:
>  Is there something we can do to build confidence it's correct, particularly as we potentially change the implementation of components()?

Tests make sure that Ord and Eq are now consistent with hashing.

> Perhaps we could instead improve the performance of next() on Components for similar effect?

I'm not sure if that's possible without regressing other things. E.g. slapping `#[inline]` on some of the functions may regress compile times on things not related to hashing.

`components` also have to operate a more complex state machine to produce the enums that it does, we're not interested in the finer distinctions here.

Some of the costs also came from hashing each component separately which is treated as a slice which the hashes the slice length which doesn't happen in the new code.